backend/src/resolvers.js

```diff
@@ -0,0 +1,73 @@
+const { UserInputError } = require("apollo-server");
+
+const resolvers = {
+    Query: {
+        users: (parent, args, context) => context.dataSources.db.allUsers(),
+        posts: (parent, args, context) => context.dataSources.db.allPosts(),
+    },
+    Mutation: {
+        write: async (parent, args, context) => {
+            const newPost = {
+                title: args.post.title,
+                authorName: args.post.author.name,
+            };
+            const author = await context.dataSources.db.getUser(newPost.authorName);
+            if (!author) {
+                throw new UserInputError("Invalid user", { invalidArgs: newPost.authorName });
+            }
+            return await context.dataSources.db.createPost(newPost);
+        },
```

Comment:
You could destructure `context` into `dataSources` right away and signal that you ignore `parent`:
```suggestion
        write: async (_, args, { dataSources }) => {
            const newPost = {
                title: args.post.title,
                authorName: args.post.author.name,
            };
            const author = await dataSources.db.getUser(newPost.authorName);
            if (!author) {
                throw new UserInputError("Invalid user", { invalidArgs: newPost.authorName });
            }
            return await dataSources.db.createPost(newPost);
        },
```

Reply:
Applies to other resolvers as well :slightly_smiling_face: